apsuite/commisslib/meas_touschek_lifetime.py

```diff
@@ -468,6 +389,301 @@ def plot_current_decay(self, fname=None, title=None):
+        gas_rate_b = 1/ltime_b - 1/tsck_b
+        anly['gas_lifetime_a'] = 1/gas_rate_a
+        anly['gas_lifetime_b'] = 1/gas_rate_b
+        anly['gas_lifetime'] = 2/(gas_rate_a + gas_rate_b)
+
+    def _calc_touschek_lifetime(self):
+        """."""
+        def model(curr, tous, blen):
+            curr_a, curr_b = curr
+            alpha_a = tous*curr_a/(1+blen*curr_a)
+            alpha_b = tous*curr_b/(1+blen*curr_b)
+            return alpha_a - alpha_b
+
+        anly = self.analysis
+        curr_a, curr_b = anly['current_a'], anly['current_b']
+        ltime_a, ltime_b = anly['total_lifetime_a'], anly['total_lifetime_b']
+        curr_a = curr_a[:ltime_a.size]
+        curr_b = curr_b[:ltime_b.size]
+
+        p0_ = (1, 1)
+        alpha_total = 1/ltime_a - 1/ltime_b
+        coeffs, pconv = _scy_opt.curve_fit(
+            model, (curr_a, curr_b), alpha_total, p0=p0_)
+        tous, blen = coeffs
+
+        alpha_a = tous*curr_a/(1+blen*curr_a)
+        alpha_b = tous*curr_b/(1+blen*curr_b)
+        anly['touschek_a'] = 1/alpha_a
+        anly['touschek_b'] = 1/alpha_b
+        anly['alpha_total'] = alpha_total
+        anly['tous_coeffs'] = coeffs
+        anly['tous_coeffs_pconv'] = pconv
+
+    def _do_measure(self):
+        meas = dict(
+            sum_a=[], sum_b=[], tim_a=[], tim_b=[], current=[],
+            rf_voltage=[], avg_pressure=[], tunex=[], tuney=[])
+        parms = self.params
+
+        curr = self.devices['currinfo']
+        rfcav = self.devices['rfcav']
+        tune = self.devices['tune']
+        press = self.pvs['avg_pressure']
+        bpm = self.devices[parms.bpm_name]
+        excx0 = self.pvs['exc_tunex_sts'].value
+        excy0 = self.pvs['exc_tuney_sts'].value
+
+        self.devices['trigger'].source = 'Study'
+        self.devices['event'].mode = 'Continuous'
+        self.devices['evg'].cmd_update_events()
+
+        self.turn_off_bpms_auto_monitor()
+        bpm.acq_nrsamples_pre = parms.acq_nrsamples_pre
+        bpm.acq_nrsamples_post = parms.acq_nrsamples_post
+        bpm.rffe_att = parms.bpm_attenuation
+        bpm.tbt_mask_enbl = 1
+        bpm.cmd_acq_abort()
+        bpm.cmd_acq_start()
+        _time.sleep(parms.wait_mask)
+
+        maxidx = parms.total_duration / (2*parms.wait_mask)
+        maxidx = float('inf') if maxidx < 1 else maxidx
+        idx = 0
+
+        if not excx0:
+            self.pvs['exc_tunex_sel'].value = 1
+        if not excy0:
+            self.pvs['exc_tuney_sel'].value = 1
+        _time.sleep(1)
+        meas['tunex'] = tune.tunex
+        meas['tuney'] = tune.tuney
+        _time.sleep(1)
+        self.pvs['exc_tunex_sel'].value = 0
+        self.pvs['exc_tuney_sel'].value = 0
+
+        while idx < maxidx and not self._stopevt.is_set():
+            bpm.tbt_mask_beg = parms.mask_beg_bunch_a
+            bpm.tbt_mask_end = parms.mask_end_bunch_a
+            _time.sleep(parms.wait_mask)
+            meas['sum_a'].append(bpm.mt_possum.mean())
```

Comment:
We should check whether the possum has `NaNs` and discard the measurement. Or we could use the `nanmean` function. Also, it would be nice to keep track of how many numbers were `NaN`s, maybe by some print message, or keeping their quantity in an array...